Firmware/Board/v3.3/Src/syscalls.c

```diff
@@ -19,6 +19,45 @@
 //int _fstat(int file, struct stat *st) {}
 //int _isatty(int file) {}
 
+extern char _end; // provided by the linker script
+extern char _heap_end_max; // provided by the linker script
+void* _end_ptr = &_end;
+void* _heap_end_max_ptr = &_heap_end_max;
+void* heap_end_ptr = 0;
+
+/* @brief Increments the program break (aka heap end)
+*
+* This is called internally by malloc once it runs out
+* of heap space. Malloc might expect a contiguous heap,
+* so we don't call the FreeRTOS pvPortMalloc here.
+* If this function returns -1, malloc will return NULL.
+* Note that if this function returns NULL, malloc does not
+* consider this as an error and will return the pointer 0x8.
+*
+* You should still be careful with using malloc though,
+* as it does not guarantee thread safety.
+*
+* @return A pointer to the newly allocated block on success
+*         or -1 otherwise.
+*/
+intptr_t _sbrk_r(struct _reent *ignored __attribute__((unused)), size_t size) {
```

Comment:
Why do we implement _sbrk_r instead of _sbrk?

Reply:
By default the call chain goes like this:
malloc => _malloc_r => _sbrk_r => _sbrk
In ODriveFirmware.elf there is no other place besides _sbrk_r that calls _sbrk directly, so currently overriding _sbrk_r has the same effect but keeps the option open to use the reent argument later.

Arguably we could also override _malloc_r such that it uses the FreeRTOS heap. The reasons why I didn't do that are:
 * I only noticed this option just now
 * If we go forward with CCM there'd me no more way to use the remaining space in main RAM